go/vumitools/middleware.py

```diff
@@ -603,6 +603,67 @@ def handle_failure(self, failure, connector_name):
         return failure
 
 
+class ConversationMetricsMiddlewareConfig(BaseMiddleware.CONFIG_CLASS):
+
+    redis_manager = ConfigDict(
+        "Redis configuration parameters", default={}, static=True)
+
+    riak_manager = ConfigRiak(
+        "Riak configuration parameters. Must contain at least a bucket_prefix"
+        " key", required=True, static=True)
+
+
+class ConversationMetricsMiddleware(BaseMiddleware):
+    """
+    Middleware that stores which conversations have received or sent messages
+
+    :param dict redis_manager:
+        Connection configuration details for Redis.
+    :param dict riak_manager:
+        Configuration details for Riak.
+    """
+
+    CONFIG_CLASS = ConversationMetricsMiddlewareConfig
+
+    @inlineCallbacks
+    def setup_middleware(self):
+        # We don't use a VumiApi here because we don't have a Riak config for
+        # it.
+        self.vumi_api = yield VumiApi.from_config_async({
+            "riak_manager": self.config.riak_manager,
+            "redis_manager": self.config.redis_manager,
+        })
+        self.redis = self.vumi_api.redis.sub_manager(
+            "conversation.metrics.middleware")
+
+    def teardown_middleware(self):
+        return self.vumi_api.close()
+
+    def record_conv_seen(self, msg):
+        mdh = MessageMetadataHelper(
+            self.vumi_api, msg)
+        conv_key = mdh.get_conversation_key()
+        acc_key = mdh.get_account_key()
+
+        # this string should be valid json
```

Comment:
Might be an idea to explain here why you're manually constructing the JSON (i.e. because we need the values added to the redis set to have exactly the value each time).